_overviews/FAQ/initialization-order.md

```diff
@@ -36,129 +36,125 @@ Let's observe the initialization order through the Scala REPL:
 Only when we get to the constructor of `C` are both `x1` and `x2` initialized. Therefore, constructors of `A` and `B` risk running into `NullPointerException`s.
 
 ## Explanation
-A 'strict' or 'eager' val is one which is not marked lazy.
 
-In the absence of "early definitions" (see below), initialization of strict vals is done in the following order.
+A "strict" or "eager" val is one which is not marked lazy.
+Initialization of strict vals is done in the following order:
 
 1. Superclasses are fully initialized before subclasses.
 2. Otherwise, in declaration order.
 
-Naturally when a val is overridden, it is not initialized more than once.  So though x2 in the above example is seemingly defined at every point, this is not the case: an overridden val will appear to be null during the construction of superclasses, as will an abstract val.
+When a `val` is overridden, in fact its accessor method (the "getter") is overridden.
+So the access to `x2` in class `A` in fact invokes the overridden getter in class `C` which reads the underlying field `C.x2`.
+This field is not yet initialized during the construction of `A`.
 
-There is a compiler flag which can be useful for identifying this situation:
+## Mitigation
 
-**-Xcheckinit**: Add runtime check to field accessors.
+The [`-Ysafe-init` compiler flag](https://docs.scala-lang.org/scala3/reference/other-new-features/safe-initialization.html) in Scala 3 enables compiler warnings for accesses to uninitialized fields:
 
-It is inadvisable to use this flag outside of testing.  It adds significantly to the code size by putting a wrapper around all potentially uninitialized field accesses: the wrapper will throw an exception rather than allow a null (or 0/false in the case of primitive types) to silently appear.  Note also that this adds a *runtime* check: it can only tell you anything about code paths which you exercise with it in place.
+    -- Warning: Test.scala:8:6 ------------------
+    8 |  val x1: String = "hello"
+      |      ^
+      |      Access non-initialized value x1. Calling trace:
+      |      ├── class B extends A {	[ Test.scala:7 ]
+      |      │   ^
+      |      ├── abstract class A {	[ Test.scala:1 ]
+      |      │   ^
+      |      └── println("A: " + x1 + ", " + x2)	[ Test.scala:5 ]
+      |                          ^^
 
-Using it on the opening example:
+In Scala 2, the `-Xcheckinit` flag adds runtime checks in the generated bytecode to identify accesses of uninitialized fields.
+The code then throws an exception rather than allowing a `null` (or `0` / `false` in the case of primitive types) to silently appear.
+Note that these runtime checks only test code that is actually exectued at runtime.
+The flag can be helpful to find accesses to uninitialized fields, but it should never be used in production due to its performance overhead.
```

Comment:
```suggestion
The flag can be helpful to find accesses to uninitialized fields, but it should not be used in production due to its performance overhead.
```

Reply:
"must never", and spoken with Germanic authority. People run debug in production all the time just in case there's a crash, which there always is.